feat: add information and examples for push downs and performance implications (#3725)
influxdata/docs-v2: content/flux/v0.x/stdlib/influxdata/influxdb/from.md

If authentication is _enabled_, provide your InfluxDB username and password
using the `<username>:<password>` syntax.

## Push down optimizations

Some transformations after the call to `from()` will trigger performance optimizations called push downs.
These optimizations are "pushed down" from Flux into the storage layer and involve utilizing some code in storage to directly apply the transformation to the data to either filter or aggregate the data.
These happen automatically and the person writing the query doesn't have to know about the technical details, but it is helpful to know about these optimizations so you can understand why certain orders of transformations may be faster than others.

Push downs require the chain between transformations to be unbroken and exclusive.
A `from()` that goes to multiple push down optimizations will cause none of them to be applied.

> To avoid this while allowing for code reuse, prefer invoking `from()` in a function when multiple push downs get applied.

Author: Much better!

### Filter

A `filter()` that makes a comparison to `r._measurement`, `r._field`, `r._value` or any tag value will be pushed down to the storage layer.
Comparisons that use functions will not.
If the function produces a static value, you can evaluate the function outside of the filter and the result will be pushed down.

> Multiple consecutive filters that would get pushed down will get merged together into a single filter that gets pushed down.

Reviewer: ~~~suggestion
`filter()` transformations that compare `r._measurement`, `r._field`, `r._value` or any tag value are pushed down to the storage layer.
Comparisons that use functions do not.
If the function produces a static value, evaluate the function outside of `filter()`.
For example:

```js\
import "strings"

// filter() is NOT pushed down
data
    |> filter(fn: (r) => r.example == strings.joinStr(arr: ["foo", "bar"], v: ""))

// filter() is pushed down
exVar = strings.joinStr(arr: ["foo", "bar"], v: ""))

data
    |> filter(fn: (r) => r.example == exVar)
```

Multiple consecutive `filter()` transformations that can be pushed down are merged together into a single filter that gets pushed down.
~~~

Reviewer: This is also true for string concatenation and interpolation, right? Filters that concat or interpolate strings can't be pushed down.

Author: I think it probably applies to interpolation and concatenation but honestly it probably shouldn't.

### Aggregates

> The following aggregates will be pushed down:

Reviewer: ```suggestion
The following aggregate transformations are pushed down:
```

- `min()`
- `max()`
- `sum()`
- `count()`
- `mean()` (except when used with `group()`)

Aggregates will also be pushed down if they are preceded by a `group()`.

> The only exception is `mean()` which cannot be pushed down to the storage layer with `group()`.

Reviewer: This list seems incomplete? It doesn't appear to include the work that Faith did and what we currently have in the documentation: https://docs.influxdata.com/influxdb/cloud/query-data/optimize-queries/#pushdown-functions-and-function-combinations

Author: Some of those functions are ones I'm not aware of being pushdowns. In particular, I've never seen rename as something that is pushed down. Duplicate is part of aggregate window so it is "pushed down" in that context, but I prefer not to over-complicate things by saying duplicate is pushed down.

I'll have to investigate. Should I adapt this document for the same page? I wasn't aware that one existed.

### Aggregate Window

> Aggregates used with `aggregateWindow()` will be pushed down.

Reviewer: ```suggestion
Aggregates used with `aggregateWindow()` are pushed down.
```

  token: token
)
```

> ### Query push downs

Reviewer: ```suggestion
### Utilize pushdowns in multiple queries
```

```js

> from(bucket: "example-bucket")
> |> range(start: -1h)
> |> filter(fn: (r) => r._measurement == "m0")
> |> filter(fn: (r) => r._field == "f0")
> |> yield(name: "filter-only")
> 
> from(bucket: "example-bucket")
> |> range(start: -1h)
> |> filter(fn: (r) => r._measurement == "m0")
> |> filter(fn: (r) => r._field == "f0")
> |> max()
> |> yield(name: "max")
> 
> from(bucket: "example-bucket")
> |> range(start: -1h)
> |> filter(fn: (r) => r._measurement == "m0")
> |> filter(fn: (r) => r._field == "f0")
> |> group(columns: ["t0"])
> |> max()
> |> yield(name: "grouped-max")
> 
> from(bucket: "example-bucket")
> |> range(start: -1h)
> |> filter(fn: (r) => r._measurement == "m0")
> |> filter(fn: (r) => r._field == "f0")
> |> aggregateWindow(every: 5m, fn: max)
> |> yield(name: "windowed-max")

Reviewer: ```suggestion
from(bucket: "example-bucket")
    |> range(start: -1h)
    |> filter(fn: (r) => r._measurement == "m0")
    |> filter(fn: (r) => r._field == "f0")
    |> yield(name: "filter-only")

from(bucket: "example-bucket")
    |> range(start: -1h)
    |> filter(fn: (r) => r._measurement == "m0")
    |> filter(fn: (r) => r._field == "f0")
    |> max()
    |> yield(name: "max")

from(bucket: "example-bucket")
    |> range(start: -1h)
    |> filter(fn: (r) => r._measurement == "m0")
    |> filter(fn: (r) => r._field == "f0")
    |> group(columns: ["t0"])
    |> max()
    |> yield(name: "grouped-max")

from(bucket: "example-bucket")
    |> range(start: -1h)
    |> filter(fn: (r) => r._measurement == "m0")
    |> filter(fn: (r) => r._field == "f0")
    |> aggregateWindow(every: 5m, fn: max)
    |> yield(name: "windowed-max")
```

```

> ### Query from the same bucket to multiple push downs

Reviewer: ```suggestion
### Query from the same bucket to multiple pushdowns
```
